firefly/kernel/intel64/cpu/cpu.cpp

```diff
@@ -3,22 +3,70 @@
 #include "firefly/intel64/cpu/ap/ap.hpp"
 #include "firefly/intel64/cpu/apic/apic.hpp"
 #include "firefly/intel64/int/interrupt.hpp"
+#include "firefly/memory-manager/allocator.hpp"
+#include "firefly/memory-manager/secondary/heap.hpp"
+#include "frg/manual_box.hpp"
+#include "frg/spinlock.hpp"
+#include "frg/vector.hpp"
 
-// Note: This is purposely bare-bones and lacking a
-// proper "per cpu" structure (i.e. gsbase).
-// There is no multi-core yet, but I'd prefer to structure
-// everything in a way so that it's easy to get it up and running now rather than later.
 namespace firefly::kernel {
-CpuData temporaryCpuInstance;
+namespace {
+frg::manual_box<CpuData> bootCpuData;
+frg::vector<CpuData *, Allocator> allCpuContexts;
+frg::ticket_spinlock allCpuContextsLock;
+}  // namespace
+
+void setupCpuContext(AssemblyCpuData *context) {
+    context->selfPointer = context;
+    wrmsr(MSR::GsBase, reinterpret_cast<uint64_t>(context));
+}
+
+void setupBootCpuContext() {
+    bootCpuData.initialize();
+    setupCpuContext(bootCpuData.get());
+}
+
+void initializeBootProccessor(uint64_t stack) {
+    setupBootCpuContext();
+    initializeThisCpu(stack);
+}
+
+void initializeApplicationProcessor(uint64_t stack) {
+    // TODO: Check with @vo1d how to do this
```

Comment:
What are you trying to check?

Reply:
It was the heap question I had some time ago

Reply:
Did we resolve that question?
I don't fully remember the question itself

Reply:
IIRC yes